README.md
scDataviz: single cell dataviz and downstream analyses
================
Kevin Blighe
2021-07-23

# Introduction

In the single cell World, which includes flow cytometry, mass cytometry,
single-cell RNA-seq (scRNA-seq), and others, there is a need to improve
data visualisation and to bring analysis capabilities to researchers
even from non-technical backgrounds. *scDataviz* (Blighe 2020) attempts
to fit into this space, while also catering for advanced users.
Additonally, due to the way that *scDataviz* is designed, which is based
on *SingleCellExperiment* (Lun and Risso 2020), it has a ‘plug and play’
feel, and immediately lends itself as flexibile and compatibile with
studies that go beyond *scDataviz*. Finally, the graphics in *scDataviz*
are generated via the *ggplot* (Wickham 2016) engine, which means that
users can ‘add on’ features to these with ease.

This package just provides some additional functions for dataviz and
clustering, and provides another way of identifying cell-types in
clusters. It is not strictly intended as a standalone analysis package.
For a comprehensive high-dimensional cytometry workflow, it is
recommended to check out the work by Nowicka *et al.* [CyTOF workflow:
differential discovery in high-throughput high-dimensional cytometry
datasets](https://f1000research.com/articles/6-748). For a more
comprehensive scRNA-seq workflow, please check out
[OSCA](https://osca.bioconductor.org) and [Analysis of single cell
RNA-seq
data](https://scrnaseq-course.cog.sanger.ac.uk/website/index.html).

# Installation

## 1\. Download the package from Bioconductor

``` r
  if (!requireNamespace('BiocManager', quietly = TRUE))
    install.packages('BiocManager')

  BiocManager::install('scDataviz')
```

Note: to install development version:

``` r
  devtools::install_github('kevinblighe/scDataviz')
```

## 2\. Load the package into R session

``` r
  library(scDataviz)
```

# Tutorial 1: CyTOF FCS data

Here, we will utilise some of the flow cytometry data from [Deep
phenotyping detects a pathological CD4+ T-cell complosome signature in
systemic sclerosis](https://www.nature.com/articles/s41423-019-0360-8).

This can normally be downloadedd via `git clone` from your command
prompt:

``` bash

  git clone https://github.com/kevinblighe/scDataviz_data/ ;
```

In a practical situation, we would normally read in this data from the
raw FCS files and then QC filter, normalise, and transform them. This
can be achieved via the `processFCS` function, which, by default, also
removes variables based on low variance and downsamples \[randomly\]
your data to 100000 variables. The user can change these via the
`downsample` and `downsampleVar` parameters. An example (*not run*) is
given below:

``` r
  filelist <- list.files(
    path = "scDataviz_data/FCS/",
    pattern = "*.fcs|*.FCS",
    full.names = TRUE)
  filelist

  metadata <- data.frame(
    sample = gsub('\\ [A-Za-z0-9]*\\.fcs$', '',
      gsub('scDataviz_data\\/FCS\\/\\/', '', filelist)),
    group = c(rep('Healthy', 7), rep('Disease', 11)),
    treatment = gsub('\\.fcs$', '',
      gsub('scDataviz_data\\/FCS\\/\\/[A-Z0-9]*\\ ', '', filelist)),
    row.names = filelist,
    stringsAsFactors = FALSE)
  metadata

  inclusions <- c('Yb171Di','Nd144Di','Nd145Di',
    'Er168Di','Tm169Di','Sm154Di','Yb173Di','Yb174Di',
    'Lu175Di','Nd143Di')

  markernames <- c('Foxp3','C3aR','CD4',
    'CD46','CD25','CD3','Granzyme B','CD55',
    'CD279','CD45RA')

  names(markernames) <- inclusions
  markernames

  exclusions <- c('Time','Event_length','BCKG190Di',
    'Center','Offset','Width','Residual')

  sce <- processFCS(
    files = filelist,
    metadata = metadata,
    transformation = TRUE,
    transFun = function (x) asinh(x),
    asinhFactor = 5,
    downsample = 10000,
    downsampleVar = 0.7,
    colsRetain = inclusions,
    colsDiscard = exclusions,
    newColnames = markernames)
```

In flow and mass cytometry, getting the correct marker names in the FCS
files can be surprisingly difficult. In many cases, from experience, a
facility may label the markers by their metals, such as Iridium (Ir),
Ruthenium (Ru), Terbium (Tb), *et cetera* - this is the case for the
data used in this tutorial. The true marker names may be held as pData
encoded within each FCS, accessible via:

``` r
  library(flowCore)
  pData(parameters(
    read.FCS(filelist[[4]], transformation = FALSE, emptyValue = FALSE)))
```

Whatever the case, it is important to sort out marker naming issues
prior to the experiment being conducted in order to avoid any confusion.

For this vignette, due to the fact that the raw FCS data is \> 500
megabytes, we will work with a smaller pre-prepared dataset that has
been downsampled to 10000 cells using the above code. This data comes
included with the package.

Load the pre-prepared complosome
data.

``` r
  load(system.file('extdata/', 'complosome.rdata', package = 'scDataviz'))
```

One can also create a new *SingleCellExperiment* object manually using
any type of data, including any data from scRNA-seq produced elsewhere.
Import functions for data deriving from other sources is covered in
Tutorials 2 and 3 in this vignette. All functions in *scDataviz*
additionally accept data-frames or matrices on their own,
de-necessitating the reliance on the *SingleCellExperiment* class.

## Perform principal component analysis (PCA)

We can use the *PCAtools* (Blighe and Lun 2020) package for the purpose
of performing PCA.

``` r
  library(PCAtools)
  p <- pca(assay(sce, 'scaled'), metadata = metadata(sce))

  biplot(p,
    x = 'PC1', y = 'PC2',
    lab = NULL,
    xlim = c(min(p$rotated[,'PC1'])-1, max(p$rotated[,'PC1'])+1),
    ylim = c(min(p$rotated[,'PC2'])-1, max(p$rotated[,'PC2'])+1),
    pointSize = 1.0,
    colby = 'treatment',
    legendPosition = 'right',
    title = 'PCA applied to CyTOF data',
    caption = paste0('10000 cells randomly selected after ',
      'having filtered for low variance'))
```

![Perform PCA](README_files/figure-gfm/ex1-1.png)

We can add the rotated component loadings as a new reduced dimensional
component to our dataset.

``` r
  reducedDim(sce, 'PCA') <- p$rotated
```

For more functionality via *PCAtools*, check the vignette: [PCAtools:
everything Principal Component
Analysis](https://bioconductor.org/packages/release/bioc/vignettes/PCAtools/inst/doc/PCAtools.html)

## Perform UMAP

UMAP can be performed on the entire dataset, if your computer’s memory
will permit. Currently it’s default is to use the data contained in the
‘scaled’ assay component of your *SingleCellExperiment* object.

``` r
  sce <- performUMAP(sce)
```

UMAP can also be stratified based on a column in your metadata, e.g.,
(treated versus untreated samples); however, to do this, I recommend
creating separate *SingleCellExperiment* objects from the very start,
i.e., from the the data input stage, and processing the data separately
for each group.

**Nota bene** - advanced users may want to change the default
configuration for UMAP. *scDataviz* currently performs UMAP via the
*umap* package. In order to modify the default configuration, one can
pull in the default config separately from the *umap* package and then
modify these config values held in the *umap.defaults* variable, as per
the [umap
vignette](https://cran.r-project.org/web/packages/umap/vignettes/umap.html)
(see ‘Tuning UMAP’ section). For example:

``` r
  config <- umap::umap.defaults
  config$min_dist <- 0.5
  performUMAP(sce, config = config)
```

We can also perform UMAP on a select number of PC eigenvectors.
*PCAtools* (Blighe and Lun 2020) can be used to infer ideal number of
dimensions to use via the elbow method and Horn’s parallel analysis.

``` r
  elbow <- findElbowPoint(p$variance)
  horn <- parallelPCA(assay(sce, 'scaled'))

  elbow
```

    ## PC3 
    ##   3

``` r
  horn$n
```

    ## [1] 1

For now, let’s just use 5 PCs.

``` r
  sce <- performUMAP(sce, reducedDim = 'PCA', dims = c(1:5))
```

## Create a contour plot of the UMAP layout

This and the remaining sections in this tutorial are about producing
great visualisations of the data and attempting to make sense of it,
while not fully overlapping with functionalioty provided by other
programs that operate in tis space.

With the contour plot, we are essentially looking at celluar density. It
can provide for a beautiful viusualisation in a manuscript while also
serving as a useful QC tool: if the density is ‘scrunched up’ into a
single area in the plot space, then there are likely issues with your
input data distribution. We want to see well-separated, high density
‘islands’, or, at least, gradual gradients that blend into one another
across high density ‘peaks’.

``` r
  ggout1 <- contourPlot(sce,
    reducedDim = 'UMAP',
    bins = 150,
    subtitle = 'UMAP performed on expression values',
    legendLabSize = 18,
    axisLabSize = 22,
    titleLabSize = 22,
    subtitleLabSize = 18,
    captionLabSize = 18)

  ggout2 <- contourPlot(sce,
    reducedDim = 'UMAP_PCA',
    bins = 150,
    subtitle = 'UMAP performed on PC eigenvectors',
    legendLabSize = 18,
    axisLabSize = 22,
    titleLabSize = 22,
    subtitleLabSize = 18,
    captionLabSize = 18)

  cowplot::plot_grid(ggout1, ggout2,
    labels = c('A','B'),
    ncol = 2, align = "l", label_size = 24)
```

![Create a contour plot of the UMAP
layout](README_files/figure-gfm/ex2-1.png)

## Show marker expression across the layout

Here, we randomly select some markers and then plot their expression
profiles across the UMAP layouts.

``` r
  markers <- sample(rownames(sce), 6)
  markers
```

    ## [1] "Foxp3"  "CD4"    "CD45RA" "CD25"   "CD279"  "CD46"

``` r
  ggout1 <- markerExpression(sce,
    markers = markers,
    subtitle = 'UMAP performed on expression values',
    nrow = 1, ncol = 6,
    legendKeyHeight = 1.0,
    legendLabSize = 18,
    stripLabSize = 22,
    axisLabSize = 22,
    titleLabSize = 22,
    subtitleLabSize = 18,
    captionLabSize = 18)

  ggout2 <- markerExpression(sce,
    markers = markers,
    reducedDim = 'UMAP_PCA',
    subtitle = 'UMAP performed on PC eigenvectors',
    nrow = 1, ncol = 6,
    col = c('white', 'darkblue'),
    legendKeyHeight = 1.0,
    legendLabSize = 18,
    stripLabSize = 22,
    axisLabSize = 22,
    titleLabSize = 22,
    subtitleLabSize = 18,
    captionLabSize = 18)

  cowplot::plot_grid(ggout1, ggout2,
    labels = c('A','B'),
    nrow = 2, align = "l", label_size = 24)
```

![Show marker expression across the
layout](README_files/figure-gfm/ex3-1.png)

## Shade cells by metadata

Shading cells by metadata can be useful for identifying any batch
effects, but also useful for visualising, e.g., differences across
treatments.

First, let’s take a look inside the metadata that we have.

``` r
  head(metadata(sce))
```

    ##       sample   group treatment
    ## cell1    P00 Disease    Unstim
    ## cell2    P00 Disease    Unstim
    ## cell3    P04 Disease      CD46
    ## cell4    P03 Disease      CD46
    ## cell5    P08 Disease    Unstim
    ## cell6    P00 Disease      CD46

``` r
  levels(metadata(sce)$group)
```

    ## [1] "Healthy" "Disease"

``` r
  levels(metadata(sce)$treatment)
```

    ## [1] "CD46"   "Unstim" "CD3"

``` r
  ggout1 <- metadataPlot(sce,
    colby = 'group',
    colkey = c(Healthy = 'royalblue', Disease = 'red2'),
    title = 'Disease status',
    subtitle = 'UMAP performed on expression values',
    legendLabSize = 16,
    axisLabSize = 20,
    titleLabSize = 20,
    subtitleLabSize = 16,
    captionLabSize = 16)

  ggout2 <- metadataPlot(sce,
    reducedDim = 'UMAP_PCA',
    colby = 'group',
    colkey = c(Healthy = 'royalblue', Disease = 'red2'),
    title = 'Disease status',
    subtitle = 'UMAP performed on PC eigenvectors',
    legendLabSize = 16,
    axisLabSize = 20,
    titleLabSize = 20,
    subtitleLabSize = 16,
    captionLabSize = 16)

  ggout3 <- metadataPlot(sce,
    colby = 'treatment',
    title = 'Treatment type',
    subtitle = 'UMAP performed on expression values',
    legendLabSize = 16,
    axisLabSize = 20,
    titleLabSize = 20,
    subtitleLabSize = 16,
    captionLabSize = 16)

  ggout4 <- metadataPlot(sce,
    reducedDim = 'UMAP_PCA',
    colby = 'treatment',
    title = 'Treatment type',
    subtitle = 'UMAP performed on PC eigenvectors',
    legendLabSize = 16,
    axisLabSize = 20,
    titleLabSize = 20,
    subtitleLabSize = 16,
    captionLabSize = 16)

  cowplot::plot_grid(ggout1, ggout3, ggout2, ggout4,
    labels = c('A','B','C','D'),
    nrow = 2, ncol = 2, align = "l", label_size = 24)
```

![Shade cells by metadata](README_files/figure-gfm/ex4-1.png)

## Find ideal clusters in the UMAP layout via k-nearest neighbours

This function utilises the k nearest neighbours (k-NN) approach from
Seurat, which works quite well on flow cytometry and CyTOF UMAP layouts,
from my experience.

``` r
  sce <- clusKNN(sce,
    k.param = 20,
    prune.SNN = 1/15,
    resolution = 0.01,
    algorithm = 2,
    verbose = FALSE)

  sce <- clusKNN(sce,
    reducedDim = 'UMAP_PCA',
    clusterAssignName = 'Cluster_PCA',
    k.param = 20,
    prune.SNN = 1/15,
    resolution = 0.01,
    algorithm = 2,
    verbose = FALSE)

  ggout1 <- plotClusters(sce,
    clusterColname = 'Cluster',
    labSize = 7.0,
    subtitle = 'UMAP performed on expression values',
    caption = paste0('Note: clusters / communities identified via',
      '\nLouvain algorithm with multilevel refinement'),
    axisLabSize = 20,
    titleLabSize = 20,
    subtitleLabSize = 16,
    captionLabSize = 16)

  ggout2 <- plotClusters(sce,
    clusterColname = 'Cluster_PCA',
    reducedDim = 'UMAP_PCA',
    labSize = 7.0,
    subtitle = 'UMAP performed on PC eigenvectors',
    caption = paste0('Note: clusters / communities identified via',
      '\nLouvain algorithm with multilevel refinement'),
    axisLabSize = 20,
    titleLabSize = 20,
    subtitleLabSize = 16,
    captionLabSize = 16)

  cowplot::plot_grid(ggout1, ggout2,
    labels = c('A','B'),
    ncol = 2, align = "l", label_size = 24)
```

![Find ideal clusters in the UMAP layout via k-nearest
neighbours](README_files/figure-gfm/ex5-1.png)

## Plot marker expression per identified cluster

``` r
  markerExpressionPerCluster(sce,
    caption = 'Cluster assignments based on UMAP performed on expression values',
    stripLabSize = 22,
    axisLabSize = 22,
    titleLabSize = 22,
    subtitleLabSize = 18,
    captionLabSize = 18)
```

``` r
  clusters <- unique(metadata(sce)[['Cluster_PCA']])
  clusters
```

    ## [1] 2 0 1 6 4 3 5 7

``` r
  markers <- sample(rownames(sce), 5)
  markers
```

    ## [1] "Foxp3"  "C3aR"   "CD25"   "CD45RA" "CD3"

``` r
  markerExpressionPerCluster(sce,
    clusters = clusters,
    clusterAssign = metadata(sce)[['Cluster_PCA']],
    markers = markers,
    nrow = 2, ncol = 5,
    caption = 'Cluster assignments based on UMAP performed on PC eigenvectors',
    stripLabSize = 22,
    axisLabSize = 22,
    titleLabSize = 22,
    subtitleLabSize = 18,
    captionLabSize = 18)
```

![Plot marker expression per identified
cluster2](README_files/figure-gfm/ex6b-1.png)

Try all markers across a single cluster:

``` r
  cluster <- sample(unique(metadata(sce)[['Cluster']]), 1)
  cluster
```

    ## [1] 1

``` r
  markerExpressionPerCluster(sce,
    clusters = cluster,
    markers = rownames(sce),
    stripLabSize = 20,
    axisLabSize = 20,
    titleLabSize = 20,
    subtitleLabSize = 14,
    captionLabSize = 12)
```

![Plot marker expression per identified
cluster3](README_files/figure-gfm/ex6c-1.png)

## Determine enriched markers in each cluster and plot the expression signature

This method also calculates metacluster abundances across a chosen
phenotype. The function returns a data-frame, which can then be exported
to do other analyses.

### Disease vs Healthy metacluster abundances

``` r
  markerEnrichment(sce,
    method = 'quantile',
    studyvarID = 'group')
```

<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">

<thead>

<tr>

<th style="text-align:right;">

Cluster

</th>

<th style="text-align:right;">

nCells

</th>

<th style="text-align:right;">

TotalCells

</th>

<th style="text-align:right;">

PercentCells

</th>

<th style="text-align:left;">

NegMarkers

</th>

<th style="text-align:left;">

PosMarkers

</th>

<th style="text-align:right;">

PerCent\_HD00

</th>

<th style="text-align:right;">

PerCent\_HD01

</th>

<th style="text-align:right;">

PerCent\_HD262

</th>

<th style="text-align:right;">

PerCent\_P00

</th>

<th style="text-align:right;">

PerCent\_P02

</th>

<th style="text-align:right;">

PerCent\_P03

</th>

<th style="text-align:right;">

PerCent\_P04

</th>

<th style="text-align:right;">

PerCent\_P08

</th>

<th style="text-align:right;">

nCell\_Healthy

</th>

<th style="text-align:right;">

nCell\_Disease

</th>

</tr>

</thead>

<tbody>

<tr>

<td style="text-align:right;">

0

</td>

<td style="text-align:right;">

3410

</td>

<td style="text-align:right;">

10000

</td>

<td style="text-align:right;">

34.10

</td>

<td style="text-align:left;">

NA

</td>

<td style="text-align:left;">

CD25+

</td>

<td style="text-align:right;">

0.0879765

</td>

<td style="text-align:right;">

9.6187683

</td>

<td style="text-align:right;">

25.483871

</td>

<td style="text-align:right;">

22.8152493

</td>

<td style="text-align:right;">

8.3870968

</td>

<td style="text-align:right;">

8.8563050

</td>

<td style="text-align:right;">

17.7126100

</td>

<td style="text-align:right;">

7.0381232

</td>

<td style="text-align:right;">

1200

</td>

<td style="text-align:right;">

2210

</td>

</tr>

<tr>

<td style="text-align:right;">

1

</td>

<td style="text-align:right;">

1928

</td>

<td style="text-align:right;">

10000

</td>

<td style="text-align:right;">

19.28

</td>

<td style="text-align:left;">

CD25-CD279-

</td>

<td style="text-align:left;">

CD3+CD45RA+

</td>

<td style="text-align:right;">

0.0000000

</td>

<td style="text-align:right;">

0.1556017

</td>

<td style="text-align:right;">

62.085062

</td>

<td style="text-align:right;">

0.5705394

</td>

<td style="text-align:right;">

0.2074689

</td>

<td style="text-align:right;">

0.0000000

</td>

<td style="text-align:right;">

0.0518672

</td>

<td style="text-align:right;">

36.9294606

</td>

<td style="text-align:right;">

1200

</td>

<td style="text-align:right;">

728

</td>

</tr>

<tr>

<td style="text-align:right;">

2

</td>

<td style="text-align:right;">

1298

</td>

<td style="text-align:right;">

10000

</td>

<td style="text-align:right;">

12.98

</td>

<td style="text-align:left;">

Granzyme B-CD279-

</td>

<td style="text-align:left;">

CD3+

</td>

<td style="text-align:right;">

15.6394453

</td>

<td style="text-align:right;">

7.8582435

</td>

<td style="text-align:right;">

2.234206

</td>

<td style="text-align:right;">

64.1756549

</td>

<td style="text-align:right;">

3.8520801

</td>

<td style="text-align:right;">

3.5439137

</td>

<td style="text-align:right;">

1.9260401

</td>

<td style="text-align:right;">

0.7704160

</td>

<td style="text-align:right;">

334

</td>

<td style="text-align:right;">

964

</td>

</tr>

<tr>

<td style="text-align:right;">

3

</td>

<td style="text-align:right;">

1236

</td>

<td style="text-align:right;">

10000

</td>

<td style="text-align:right;">

12.36

</td>

<td style="text-align:left;">

NA

</td>

<td style="text-align:left;">

CD25+

</td>

<td style="text-align:right;">

6.3915858

</td>

<td style="text-align:right;">

1.1326861

</td>

<td style="text-align:right;">

1.375404

</td>

<td style="text-align:right;">

2.9126214

</td>

<td style="text-align:right;">

24.0291262

</td>

<td style="text-align:right;">

7.0388350

</td>

<td style="text-align:right;">

56.8770227

</td>

<td style="text-align:right;">

0.2427184

</td>

<td style="text-align:right;">

110

</td>

<td style="text-align:right;">

1126

</td>

</tr>

<tr>

<td style="text-align:right;">

4

</td>

<td style="text-align:right;">

962

</td>

<td style="text-align:right;">

10000

</td>

<td style="text-align:right;">

9.62

</td>

<td style="text-align:left;">

NA

</td>

<td style="text-align:left;">

CD3+

</td>

<td style="text-align:right;">

16.5280665

</td>

<td style="text-align:right;">

26.1954262

</td>

<td style="text-align:right;">

7.484407

</td>

<td style="text-align:right;">

27.9625780

</td>

<td style="text-align:right;">

3.2224532

</td>

<td style="text-align:right;">

10.0831601

</td>

<td style="text-align:right;">

6.0291060

</td>

<td style="text-align:right;">

2.4948025

</td>

<td style="text-align:right;">

483

</td>

<td style="text-align:right;">

479

</td>

</tr>

<tr>

<td style="text-align:right;">

5

</td>

<td style="text-align:right;">

502

</td>

<td style="text-align:right;">

10000

</td>

<td style="text-align:right;">

5.02

</td>

<td style="text-align:left;">

CD25-

</td>

<td style="text-align:left;">

NA

</td>

<td style="text-align:right;">

0.3984064

</td>

<td style="text-align:right;">

14.9402390

</td>

<td style="text-align:right;">

40.836653

</td>

<td style="text-align:right;">

11.1553785

</td>

<td style="text-align:right;">

0.0000000

</td>

<td style="text-align:right;">

0.3984064

</td>

<td style="text-align:right;">

0.1992032

</td>

<td style="text-align:right;">

32.0717131

</td>

<td style="text-align:right;">

282

</td>

<td style="text-align:right;">

220

</td>

</tr>

<tr>

<td style="text-align:right;">

6

</td>

<td style="text-align:right;">

300

</td>

<td style="text-align:right;">

10000

</td>

<td style="text-align:right;">

3.00

</td>

<td style="text-align:left;">

CD46-CD279-

</td>

<td style="text-align:left;">

NA

</td>

<td style="text-align:right;">

0.0000000

</td>

<td style="text-align:right;">

41.0000000

</td>

<td style="text-align:right;">

2.000000

</td>

<td style="text-align:right;">

54.6666667

</td>

<td style="text-align:right;">

0.3333333

</td>

<td style="text-align:right;">

0.3333333

</td>

<td style="text-align:right;">

0.6666667

</td>

<td style="text-align:right;">

1.0000000

</td>

<td style="text-align:right;">

129

</td>

<td style="text-align:right;">

171

</td>

</tr>

<tr>

<td style="text-align:right;">

7

</td>

<td style="text-align:right;">

281

</td>

<td style="text-align:right;">

10000

</td>

<td style="text-align:right;">

2.81

</td>

<td style="text-align:left;">

NA

</td>

<td style="text-align:left;">

Foxp3+CD25+CD3+

</td>

<td style="text-align:right;">

0.3558719

</td>

<td style="text-align:right;">

0.0000000

</td>

<td style="text-align:right;">

70.462633

</td>

<td style="text-align:right;">

2.1352313

</td>

<td style="text-align:right;">

0.7117438

</td>

<td style="text-align:right;">

0.7117438

</td>

<td style="text-align:right;">

0.3558719

</td>

<td style="text-align:right;">

25.2669039

</td>

<td style="text-align:right;">

199

</td>

<td style="text-align:right;">

82

</td>

</tr>

<tr>

<td style="text-align:right;">

8

</td>

<td style="text-align:right;">

61

</td>

<td style="text-align:right;">

10000

</td>

<td style="text-align:right;">

0.61

</td>

<td style="text-align:left;">

CD46-

</td>

<td style="text-align:left;">

NA

</td>

<td style="text-align:right;">

0.0000000

</td>

<td style="text-align:right;">

18.0327869

</td>

<td style="text-align:right;">

1.639344

</td>

<td style="text-align:right;">

0.0000000

</td>

<td style="text-align:right;">

1.6393443

</td>

<td style="text-align:right;">

78.6885246

</td>

<td style="text-align:right;">

0.0000000

</td>

<td style="text-align:right;">

0.0000000

</td>

<td style="text-align:right;">

12

</td>

<td style="text-align:right;">

49

</td>

</tr>

<tr>

<td style="text-align:right;">

9

</td>

<td style="text-align:right;">

22

</td>

<td style="text-align:right;">

10000

</td>

<td style="text-align:right;">

0.22

</td>

<td style="text-align:left;">

CD46-CD279-

</td>

<td style="text-align:left;">

CD3+

</td>

<td style="text-align:right;">

0.0000000

</td>

<td style="text-align:right;">

18.1818182

</td>

<td style="text-align:right;">

4.545454

</td>

<td style="text-align:right;">

0.0000000

</td>

<td style="text-align:right;">

0.0000000

inst/NEWS

  CHANGES IN VERSION 1.2.0
  ------------------------
    - Tidied labeling in various functions when drawConnectors = TRUE.
      This is to adjust to changes in ggrepel.
    - Removed labhjust and labvjust from all functions

  CHANGES IN VERSION 1.0.0
  ------------------------
    - Submitted to Bioconductor


R/processFCS.R
#' Input, filter, normalise, and transform FCS expression data.
#'
#' @param files A vector of FCS files.
#' @param assayname Name of the assay slot in which data will be stored.
#' @param metadata Metadata associated with the FCS files specified in
#'   'files'. A strict rule is enforced requiring that \code{rownames(metadata)}
#'   matches files in both name and order.
#' @param filter Boolean (TRUE / FALSE) to enable filtering (per sample)
#'   for background signal / noise.
#' @param bgNoiseThreshold Threshold for background noise. Used when 
#'   \code{filter == TRUE}.
#' @param euclideanNormThreshold Euclidean norm threshold for background
#'   noise. Used when \code{filter == TRUE}.
#' @param transformation Boolean (TRUE / FALSE) to enable data transformation
#'   after filtering.
#' @param transFun The function to apply (per sample) for transformation. 
#'   Typically, for flow and mass cytometry, this is hyperbolic arc sine
#'   (\code{asinh(x)}). User can supply any function.
#' @param asinhFactor The factor to apply when transforming via \code{asinh()}. For
#'   flow cytometry, this is usually 150; for mass cytometry and CyTOF, it is
#'   5. Note that this is not used if the user has supplied their own function
#'   to \code{transFun}.
#' @param downsample Downsample to this number of random variables. This is
#'   perfromed on the final merged dataset, i.e., after all samples have been
#'   bound together. NULL to disable.
#' @param downsampleVar Downsample based on variance. Removes this proportion of
#'   cells based on lesser variance. This is applied per sample. If user wishes
#'   to apply this globally on the final merged dataset, then set this to 0 and
#'   remove based on variance manually.
#' @param colsDiscard Columns to be removed from the final merged data. These
#'   names are literal and must match exactly.
#' @param colsRetain Retain these columns only. This is the same as \code{colsDiscard}
#'   but in reverse. Technically, it is possible to activate both \code{colsDiscard}
#'   and \code{colsRetain}, but \code{colsDiscard} will be executed first.
#' @param newColnames A named vector of new marker names to assign to each sample.
#'   The values of this vector should be the new marker names; the names of this
#'   vector should represent the ooriginal marker names. This operation is performed
#'   AFTER any operation involving \code{colsDiscard} and \code{colsRetain}.
#' @param emptyValue boolean (taken from ?flowCore::read.FCS indicating whether or
#'   not we allow an empty value for keyword values in TEXT segment.
#' @param verbose Boolean (TRUE / FALSE) to print messages to console or not.
#'
#' @details
#' Input, filter, normalise, and transform FCS expression data.
#'
#' @return A \code{SingleCellExperiment} object.
#'
#' @author Kevin Blighe <kevin@clinicalbioinformatics.co.uk>
#'
#' @examples
#' # create random data that follows a negative binomial
#' mat1 <- jitter(matrix(
#'   MASS::rnegbin(rexp(50000, rate=.1), theta = 4.5),
#'   ncol = 20))
#' colnames(mat1) <- paste0('CD', 1:ncol(mat1))
#' rownames(mat1) <- paste0('cell', 1:nrow(mat1))
#'
#' mat2 <- jitter(matrix(
#'   MASS::rnegbin(rexp(50000, rate=.1), theta = 4.5),
#'   ncol = 20))
#' colnames(mat2) <- paste0('CD', 1:ncol(mat2))
#' rownames(mat2) <- paste0('cell', 1:nrow(mat2))
#'
#' metadata <- data.frame(
#'   group = c('PB1', 'PB2'),
#'   row.names = c('mat1', 'mat2'),
#'   stringsAsFactors = FALSE)
#'
#' @import SingleCellExperiment
#' 
#' @importFrom MASS rnegbin
#' @importFrom flowCore read.FCS exprs
#' @importFrom S4Vectors metadata<-
#'
#' @export
processFCS <- function(
  files,
  assayname = 'scaled',
  metadata = NULL,
  filter = TRUE,
  bgNoiseThreshold = 1,
  euclideanNormThreshold = 1,
  transformation = TRUE,
  transFun = function (x) asinh(x),
  asinhFactor = 5,
  downsample = 100000,
  downsampleVar = 0.1,
  colsDiscard = NULL,
  colsRetain = NULL,
  newColnames = NULL,
  emptyValue = TRUE,
  verbose = TRUE)
{

  # if metadata specified, enforce rule that rownames(metadata) is the
  # same as files
  if (!is.null(metadata)) {
    if(!identical(files, rownames(metadata))) {
      stop("'files' is not identical to 'rownames(metadata)'")
    }
  }

  # is metadata NULL?
  if (is.null(metadata)) {
    if (verbose) warning('No metadata detected - creating generic metadata')
    metadata <- data.frame(
      sample1 = files,
      sample2 = files,
      row.names = files)
  }

  # read in the data to a list
  samples <- list()
  samples <- lapply(files,
    function(x) exprs(read.FCS(x, transformation = FALSE, emptyValue = emptyValue)))
  names(samples) <- files

  # filter markers out
  if (!is.null(colsDiscard)) {
    samples <- lapply(
      samples,
      function(x) if (length(which(colnames(x) %in% colsDiscard))) {
        x[,-which(colnames(x) %in% colsDiscard)]} else {return(x)})
  }

  # filter markers in
  if (!is.null(colsRetain)) {
    samples <- lapply(
      samples,
      function(x) if (length(which(colnames(x) %in% colsRetain))) {
        x[,which(colnames(x) %in% colsRetain)]} else {return(x)})
  }

  # ensure sort order before procedding to next steps
  # this will avoid numerous issues:
  #   1. avoid situations where one marker may not be screen by all samples
  #   2. ensure that new marker names that may be assigned are assigned
  #      faithfully
  uniquenames <- sort(unique(unlist(lapply(samples, colnames))))
  if (verbose) message('--harmonising markers across samples')
  for(i in seq_len(length(samples))) {
      samples[[i]] <- samples[[i]][,match(uniquenames, colnames(samples[[i]]))]
  }

  # rename markers
  if(!is.null(newColnames)) {
    for(i in seq_len(length(samples))) {
      samples[[i]] <- samples[[i]][,match(names(newColnames), colnames(samples[[i]]))]
      colnames(samples[[i]]) <- newColnames
    }
  }

  # filter
  if (filter) {
    if (verbose) message('--filtering background / noise')

    # Euclidean norm
    samples <- lapply(
      samples,
      function(x)
        x[apply(x, 1, FUN = function(x) sqrt(sum(x^2))) > euclideanNormThreshold,])

    # noise correction
    for(i in seq_len(length(samples))) {
      x <- samples[[i]]
      x[x < bgNoiseThreshold] <- 0
      samples[[i]] <- x
    }
  }

  # transform
  if (transformation) {
    if (verbose) message('--transforming data')
    samples <- lapply(
      samples,
      function(x) transFun(x / asinhFactor))
  }

  # load function for downsampling based on variance
  if(!is.null(downsampleVar)) {
    if (downsampleVar > 0) {
      if (verbose) message('--removing the lower ', downsampleVar * 100,
        '% of cells based on variance')

      samples <- lapply(
        samples,
        function(x) downsampleByVar(x, varianceFactor = downsampleVar,
          verbose = verbose))
    }
  }

  # expand metadata to align with expression data
  names <- colnames(metadata)
  metanew <- list()
  for (i in seq(length(samples))) {
    tmp <- data.frame(row.names = seq_len(nrow(samples[[i]])))
    for (j in seq_len(ncol(metadata))) {
      tmp <- cbind(tmp, rep(metadata[i,j], nrow(samples[[i]])))
    }
    metanew[[i]] <- tmp
  }

  metadata <- do.call(rbind, metanew)
  colnames(metadata) <- names
  rownames(metadata) <- paste0('cell', seq_len(nrow(metadata)))

  # combine all samples
  samples <- do.call(rbind, samples)
  rownames(samples) <- paste0('cell', seq_len(nrow(samples)))

  # downsample
  if (!is.null(downsample)) {
    if (downsample > nrow(samples)) {
      warning('Cannot downsample to ', downsample, ' number of variables as',
        ' there are ', nrow(samples), ' variables currently in the merged ',
        'dataset.')
      if (verbose) message('--skipping downsampling')
    } else {
      if (verbose) message('--downsampling to ', downsample, ' variables.')
      idx <- sample(seq(nrow(samples)), downsample)
      samples <- samples[idx,]
      metadata <- metadata[idx,]

      rownames(metadata) <- paste0('cell', seq_len(nrow(metadata)))
      rownames(samples) <- paste0('cell', seq_len(nrow(samples)))
    }
  }

  # these should be equal
  if (!is.null(metadata)) {
    if (nrow(metadata) != nrow(samples)) {
      stop(paste0('Metadata does not match expression data',
        ' - please check your input.'))
    }
  }

  # return a SingleCellExperiment object
  ret <- list(t(samples))
  names(ret)[1] <- assayname
  ret <- SingleCellExperiment(
    assays = ret)
  metadata(ret) <- metadata
  return(ret)
}
